lib/charms/layer/apache_zeppelin.py

```diff
@@ -13,15 +24,37 @@ class Zeppelin(object):
 
     :param DistConfig dist_config: The configuration container object needed.
     """
+    @classmethod
+    def get(cls):
+        """
+        Dynamically instantiate this or a subclass, to allow other layers
+        to override portions of this implementation.
+        """
+        impl = layer.options('apache-zeppelin')['implementation_class']
+        module_name, cls_name = impl.rsplit('.', 1)
+        cls = getattr(import_module(module_name), cls_name)
+        return cls()
```

Comment:
I went with the getter to make it clear that it's a factory method.  I kind of like the `__new__` idea, but it feels a bit subtle and prone to surprise.

Another option would be to make charms.layer.apache_zeppelin a submodule with an `__init__.py` and `implementation.py` or similar, and then the layer could override the import in `__init__.py` to set the subclass to use.

However, I think a best approach would be to do more refactoring to eliminate the need for the subclass entirely.  /me ponders.